controllers/postgresql_controller.go

```diff
@@ -94,3 +100,59 @@ func (a *postgreSQLAdapter) getServiceType() string {
 func (a *postgreSQLAdapter) getDiskSpace() string {
 	return a.Spec.DiskSpace
 }
+
+func (a *postgreSQLAdapter) performUpgradeTaskIfNeeded(ctx context.Context, avn *aiven.Client, avnGen avngen.Client, old *service.ServiceGetOut) error {
+	var currentVersion string = old.UserConfig["pg_version"].(string)
+	var targetVersion string = *a.getUserConfig().(pguserconfig.PgUserConfig).PgVersion
+
+	// No need to upgrade if pg_version hasn't changed
+	if targetVersion == currentVersion {
+		return nil
+	}
+
+	task, err := avnGen.ServiceTaskCreate(ctx, a.getServiceCommonSpec().Project, a.getObjectMeta().Name, &service.ServiceTaskCreateIn{
+		TargetVersion: service.TargetVersionType(targetVersion),
+		TaskType:      service.TaskTypeUpgradeCheck,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot create PG upgrade check task: %q", err)
+	}
+
+	didTaskSucceed, err := waitForTaskToComplete(ctx, func() (*bool, error) {
+		t, getErr := avn.ServiceTask.Get(ctx, a.getServiceCommonSpec().Project, a.getObjectMeta().Name, task.TaskId)
+		if getErr != nil {
+			return nil, fmt.Errorf("error fetching service task %s: %q", t.Task.Id, getErr)
+		}
+
+		if t.Task.Success == nil {
+			return nil, nil
+		}
+		return t.Task.Success, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	if !*didTaskSucceed {
+		return fmt.Errorf("PG service upgrade check error, version upgrade from %s to %s, result: %s", currentVersion, targetVersion, task.Result)
+	}
+	return nil
+}
+
+func waitForTaskToComplete[T any](ctx context.Context, f func() (*T, error)) (*T, error) {
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context timeout while retrying operation, error=%q", err)
+		case <-time.After(waitForTaskToCompleteInterval):
+			val, err := f()
+			if err != nil {
+				return nil, err
+			}
+			if val != nil {
```

Comment:
OK, val can be nil (means task is still in progress). But what we really want to know if it is ready. So combining with this https://github.com/aiven/aiven-operator/pull/645/files#r1500504286

```suggestion
			if val {
				return true, nil
			}
```